Adding functionality to compute all_worm prevalence for SCH

adiramani opened the pull request:
- Allowing for the calculation of `all_worm` prevalence for SCH with the following caveats
 - the "first worm" is always haematobium, and doesn't have high or low prevalences
 - if the IU exists in haematobium, it should exist in the other worm's folders once
   - other worms are split by high and low burden. The IU should only appear once, in either high or in low.
- If an IU has a data file in the second worm, but not in the first worm, we do not process it (@thk123 can you confirm if this is the functionality we want)
 


Todo: Unit tests for functions/functionality added

thk123 commented:
This seems like it does what is required, thanks! On the question - that's fine but it would be good to raise a warning (I don't think there is a good reason for not having a 1:1 mapping between the two worms, so silently skipping probably isn't a good idea.

tlestang commented:
Hi Adi,

I've been looking at the postprocessing repo this morning. I feel I made progress in understanding what is going on, but I'm
afraid I won't be able to provide a full review yet.

I started by looking at `run_sch.canonicalise_raw_sth_results` since it looks simpler than the corresponding sth function.  If I
understand correctly, it takes list of worm directory names and build dataframe that is the combination of individual result
dataframes from each worm directory.  Providing a single worm dir is not a special case, since this will simply lead to
`other_worms_canonical` to be the empty list.

Because each worm directory tree has the same structure, this can be implemented by looping through the files within the first
worm directory tree, and process (`canonicalise_single_result`) all other twin files within the other worm directories.

Fine. However I don't understand why the interface (and implementation) for `canonicalise_raw_sch_results` must be different.  In
this case the function is passed a `all_worm` boolean and a `first_worm` string.

If `input_dir` is the path the the directory containing the worm directories, then I don't get why these two extra (with respect
to the sth twin function) parameters are required.  In other words, is there a reason why  `canonicalise_raw_sch_results` cannot be implemented in a way similar to:

```python
def canonicalise_raw_sch_results(
    input_dir,
    output_dir,
    worm_directories = [],
):
    # Assuming that the first worm only has one burden
    file_iter_addition = worm_directories[0]
    file_iter = get_sch_flat(f"{input_dir}{file_iter_addition}")

    all_files = list(file_iter)

    if len(all_files) == 0:
        raise Exception(
            "No data for IUs found - see above warnings and check input directory"
        )

    other_worm_directories = worm_directories[1:]
    iter_other_worms = [
        list(get_sch_flat(f"{input_dir}{other_worm_dir}"))
        for other_worm_dir in other_worm_directories
    ]

    # ...

    for file_info in tqdm(all_files, desc="Canoncialise SCH results"):
        other_worm_file_infos = []
        for worm, burden, _, _ in other_worms:
            # ...
	    # ...
        canonical_result_first_worm = canoncialise_single_result(file_info)

        # The empty list if 'worm_directories' is a list of a single
        # directory path
        other_worms_canoncial = [
            canoncialise_single_result(other_worm_file_info)
            for other_worm_file_info in other_worm_file_infos
        ]

        # If 'other_worms_canonical' is the empty list then
        # 'all_worms_canonical' is 'canonical_result_first_worm'
        all_worms_canonical = combine_many_worms(
            canonical_result_first_worm, other_worms_canoncial
        )
        output_directory_structure.write_canonical(
            output_dir, file_info, all_worms_canonical
        )
```

tlestang commented on endgame_postprocessing/model_wrappers/sch/run_sch.py:
```diff
@@ -153,9 +164,85 @@ def canonicalise_raw_sth_results(input_dir, output_dir, worm_directories):
+def canonicalise_raw_sch_results(
+    input_dir,
+    output_dir,
+    all_worm=False,
```
I read "any worm" in the `model_wrappers/sch/README.md`, but `all_worm` here. I think this is referring to the same thing? Also it looks to me that `all_worm = True` could mean only processing a subset of worm dirs, not necessarily *all* of them, correct?

Ah - and I think the README file mentionned above should be updated to describe the new behaviour you implement in this PR.

adiramani replied:
Good points for the naming, and updating the README! I shall update those!

tlestang commented on endgame_postprocessing/model_wrappers/sch/run_sch.py:
```diff
@@ -193,56 +193,38 @@ def _check_iu_in_all_folders(worm_iu_info):
+        file.file_path
+        for worm_list in iter_other_worms
+        for file in worm_list
+    ]
```
About the two above statements:
- I would move them closer to the line `flattened_iter_other_worms` is actually used, which is quite further down at line 249.
- I would also make use of the fact that `get_sch_flat`returns a generator function to avoid converting to a list and compute a set of all the file paths directly:
    ```python
    all_other_worms_file_paths = set(
        file_desc.file_path
        for worm_files_list in (
            get_sch_flat(f"{input_dir}{worm_dir}")
            for worm_dir in other_worm_directories
         )
         for file_desc in worm_files_list
    )
    for worm, burden, _, _ in other_worms:
        # ...
        if new_file.file_path in all_other_worms_file_paths:
            other_worm_file_infos.append(new_file)
    ```

In addition I renamed `file` to `file_desc` since `file` is a reserved Python keyword.

tlestang commented on endgame_postprocessing/model_wrappers/sch/run_sch.py:
```diff
@@ -252,34 +234,34 @@ def canonicalise_raw_sch_results(
+        #     output_directory_structure.write_canonical(
+        #         output_dir, file_info, canonical_result
+        #     )
+        # else:
```
Woops some dead code made it through :)

tlestang commented:
Thanks for updating the code @adiramani , I think it's a lot more consistent with the lf wrapper code now. 

Could you update the README as well to document the sch "any worm" behaviour? What is the best terminology? "any worm"? "all worms"? "multiple worms"?

adiramani commented:
> Thanks for updating the code @adiramani , I think it's a lot more consistent with the lf wrapper code now.
> 
> Could you update the README as well to document the sch "any worm" behaviour? What is the best terminology? "any worm"? "all worms"? "multiple worms"?

Thanks for the comments! I've updated the code to address them, and I've also updated the readme.

I believe "any worm" is the best terminology. From what I could gather from Thomas's notes, we are calculating the chance that an individual has any one of haematobium or mansoni.
